src/conn/conn_open.c

```diff
@@ -198,33 +198,42 @@ __wt_connection_close(WT_CONNECTION_IMPL *conn)
 int
 __wt_connection_workers(WT_SESSION_IMPL *session, const char *cfg[])
 {
+    WT_DECL_RET;
+
     /*
      * Start the optional statistics thread. Start statistics first so that other optional threads
      * can know if statistics are enabled or not.
      */
     WT_RET(__wt_statlog_create(session, cfg));
     WT_RET(__wt_logmgr_create(session, cfg));
 
+    /* Initialize metadata tracking, required before creating tables. */
+    WT_RET(__wt_meta_track_init(session));
+
+    /* Try to create the history store table. */
+    ret = __wt_hs_create(session, cfg);
+
+    if (ret != 0 && ret != ENOENT && ret != WT_NOTFOUND)
+        return (ret);
+
     /*
      * Run recovery. NOTE: This call will start (and stop) eviction if recovery is required.
      * Recovery must run before the history store table is created (because recovery will update the
      * metadata), and before eviction is started for real.
      */
     WT_RET(__wt_txn_recover(session));
 
+    /* Try to create the history store table if we failed to create it before recovery. */
```

Comment:
The `ret` value is already replaced with the return value from `__wt_txn_recover()` function.  And also can you provide the details of why it can fail before recovery.

Reply:
WT_RET doesn't overwrite `ret`. It just assigns the return value to another local variable `__ret`. @luke-pearson Can you explain why it can fail before recovery? Feel free to directly amend the comment.

Reply:
See updates on ticket, if anything isn't clear please ask for clarification.